src/resolver.ts

```diff
@@ -190,19 +189,22 @@ export class Resolver {
     this.cache = cache === true ? inMemoryCache() : cache || noCache
   }
 
-  async resolve(didUrl: string): Promise<DIDDocument> {
+  async resolve(
+    didUrl: string,
+    options: DIDResolutionOptions = {}
+  ): Promise<DIDResolutionResult> {
     const parsed = parse(didUrl)
-    const resolver = this.registry[parsed.method]
-    if (resolver) {
-      const doc = await this.cache(parsed, () =>
-        resolver(parsed.did, parsed, this)
-      )
-      if (doc == null) {
-        throw new Error(`resolver returned null for ${parsed.did}`)
-      } else {
-        return doc
+    if (parsed === null) {
+      return {
+        didResolutionMetadata: { error: 'invalidDid' },
+        didDocument: null,
+        didDocumentMetadata: {}
       }
     }
-    throw new Error(`Unsupported DID method: '${parsed.method}'`)
+    const resolver = this.registry[parsed.method]
+    if (!resolver) {
+      throw new Error(`Unsupported DID method: '${parsed.method}'`)
+    }
+    return this.cache(parsed, () => resolver(parsed.did, parsed, this, options))
```

Comment:
However, the old `DIDDocument` type is not compatible with the new one and would need to be converted somehow. Also all baseXX encoding are not supported anymore, so we would need to do some conversion there.

Reply:
right, but there is no check here either for the correct `DIDDocument` format coming from the method implementations. the result is simply bubbled up.
The only check is done by the IDE/transpiler in case the method implementation is also using typescript, but that is not really a guarantee.

Reply:
Sure, but what are you suggesting instead?

Reply:
To inspect the result and wrap it if it's legacy:

```suggestion
    return this.cache(parsed, () => {
      let result = null
      try {
        result = await resolver(parsed.did, parsed, this, options)
      } catch (e) {
        return {
          didResolutionMetadata: {
            error: 'notFound' // or perhaps another error code
            message: e.toString() // This is not in spec, nut may be helpful
          },
          didDocument: null,
          didDocumentMetadata: {}
        }
      }
      if (!result.hasOwnProperty('didResolutionMetadata') 
          && !result.hasOwnProperty('didDocumentMetadata')
          && !result.hasOwnProperty('didDocument')) {
        return {didResolutionMetadata: {}, didDocument: result, didDocumentMetadata: {}}
      } else {
        return result
      }
    })
```

Reply:
Don't like this because it will confuse the method signature of the DIDResolver to return either DIDResolutionResult or DIDDocument.